src/js.c

```diff
@@ -2827,6 +2903,219 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+    if (s == NATS_OK)
+    {
+        fetch->requestedMsgs += req.Batch;
+    }
+    natsSub_Unlock(sub);
+
+    natsBuf_Destroy(&buf);
+    return NATS_UPDATE_ERR_STACK(s);
+}
+
+static bool _autoNextFetchRequest(jsFetchRequest *req, natsSubscription *sub, void *closure)
+{
+    jsFetch *fetch = (jsFetch *)closure;
+
+    int remainingUnrequested = 0;
+    int64_t remainingBytes = 0;
+    int want = 0;
+    bool maybeMore = true;
+
+    natsSub_Lock(sub);
+
+    int isAhead = fetch->requestedMsgs - fetch->deliveredMsgs;
+    int wantAhead = fetch->keepAhead;
+    if (isAhead > wantAhead)
+        maybeMore = false;
+
+    if (maybeMore)
+    {
+        // fetch->lifetime.Batch is always > 0
+        remainingUnrequested = fetch->lifetime.Batch - fetch->requestedMsgs;
+        if (remainingUnrequested <= 0)
+            maybeMore = false;
+    }
+
+    if (maybeMore && (fetch->lifetime.MaxBytes > 0))
+    {
+        remainingBytes = fetch->lifetime.MaxBytes - fetch->receivedBytes;
+        if (remainingBytes <= 0)
+            maybeMore = false;
+    }
+
+    if (maybeMore)
+    {
+        want = remainingUnrequested;
+        if (want > fetch->fetchSize)
+            want = fetch->fetchSize;
+
+        maybeMore = (want > 0);
+    }
+
+    natsSub_Unlock(sub);
+
+    if (!maybeMore)
+        return false;
+
+    *req = fetch->lifetime; // copy bytes, reading immutable data
+    req->Batch = want;
+    // FIXME discuss in PR - this seems wrong, we don't know how many bytes we will have
+    // received from what is already requested. Still, can serve as a safe
+    // upper boundary.
+    req->MaxBytes = remainingBytes;
+    return true;
+}
+
+natsStatus
+js_PullSubscribeAsync(natsSubscription **newsub, jsCtx *js, const char *subject, const char *durable,
+                      natsMsgHandler msgCB, void *msgCBClosure,
+                      jsFetchRequest *lifetime,
+                      jsOptions *jsOpts, jsSubOptions *opts, jsErrCode *errCode)
+{
+    natsStatus s = NATS_OK;
+    natsSubscription *sub = NULL;
+    jsSub *jsi = NULL;
+    jsFetch *fetch = NULL;
+    jsFetchRequest defaultLifetime = {
+        .Batch = INT_MAX,     // no limit
+        .Expires = INT64_MAX, // never
+        .Heartbeat = 0,       // none
+        .MaxBytes = 0,        // no limit
+        .NoWait = false,      // wait forever
+    };
+
+    if ((newsub == NULL) || (msgCB == NULL))
+        return nats_setDefaultError(NATS_INVALID_ARG);
+
+    if (errCode != NULL)
+        *errCode = 0;
+
+    if (lifetime == NULL)
+        lifetime = &defaultLifetime;
+
+    // Do a basic pull subscribe first, but with a callback so it is treated as
+    // "async" and assigned to a dispatcher. Since we don't fetch anything, it
+    // will not be active yet.
+    s = _subscribe(&sub, js, subject, durable, msgCB, msgCBClosure, true, jsOpts, opts, errCode);
+
+    IFOK(s, (fetch = NATS_CALLOC(1, sizeof(jsFetch))) == NULL ? NATS_NO_MEMORY : NATS_OK);
```

Comment:
Inconvenient of that is that you are not calling `nats_setDefaultError(NATS_NO_MEMORY)`. So even though it is more code, you should have:
```
if (s == NATS_OK)
{
    fetch = NATS_CALLOC(1, sizeof(jsFetch));
    if (fetch == NULL)
        s = nats_setDefaultError(NATS_NO_MEMORY);
}
```